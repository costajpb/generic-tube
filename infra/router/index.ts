import { createRouter, createWebHistory } from 'vue-router'
import DashboardView from '@/infra/views/DashboardView.vue'

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: '/',
      name: 'dashboard',
      component: DashboardView
    },
    {
      path: '/shows/:id',
      name: 'show-details',
      component: () => import('@/infra/views/ShowDetailsView.vue')
    }
  ]
})

export default router
