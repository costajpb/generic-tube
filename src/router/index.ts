import { createRouter, createWebHistory } from 'vue-router'
import DashboardView from '@/src/views/DashboardView.vue'

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: '/',
      name: 'dashboard',
      component: DashboardView
    },
    {
      path: '/shows/:id',
      name: 'show-details',
      component: () => import('@/src/views/ShowDetailsView.vue')
    }
  ]
})

export default router
