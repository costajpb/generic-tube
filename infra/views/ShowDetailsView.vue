<script setup lang="ts">
    import { provide } from 'vue';
    import ShowDetails from '@/infra/adapters/use-cases/show-details'
    import { useRoute } from 'vue-router';
    import PageDetails from '@/infra/components/PageDetails.vue';
    import router from '@/infra/router';

    const route = useRoute()
    const useCase = new ShowDetails(parseInt(route.params.id as string))
    const details = useCase.details
    
    useCase.on('showDetails:return', () => {
        router.push('/')
    })
    
    provide('useCase', useCase)
</script>

<template>
    <PageDetails :details="details" v-if="details" />
</template>