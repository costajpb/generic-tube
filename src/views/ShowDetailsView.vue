<script setup lang="ts">
    import { onMounted, provide, ref } from 'vue';
    import ShowDetails from '@/src/adapters/use-cases/show-details'
    import { useRoute } from 'vue-router';
    import type Show from '@/domain/show/entity';
    import ShowDetailsPage from '@/src/components/ShowDetailsPage.vue';
    import router from '../router';

    const details = ref<Required<Pick<Show, 'episodes'>> & Show | undefined>(undefined)
    const route = useRoute()
    const useCase = new ShowDetails(parseInt(route.params.id as string))
    
    useCase.on('showDetails:return', () => {
        router.push('/')
    })
    
    provide('useCase', useCase)

    onMounted(async () => {
        details.value = (await useCase.details) as Required<Pick<Show, 'episodes'>> & Show
    })
</script>

<template>
    <ShowDetailsPage :details="details" v-if="details" />
</template>