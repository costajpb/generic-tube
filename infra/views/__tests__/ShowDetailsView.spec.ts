import router from "@/infra/router"
import { flushPromises, mount } from "@vue/test-utils"
import ShowDetailsView from "@/infra/views/ShowDetailsView.vue"
import ShowDetails from '@/application/show-details'

vi.mock('vue-router', async (importOriginal) => {
    const mod = await importOriginal<typeof import('vue-router')>()
    return {
        ...mod,
        useRoute: () => ({
            params: {
                id: '1'
            }
        })
    }
})

describe('ShowDetailsView', () => {
    it('should return to dashboard upon request', async () => {
        vi.spyOn(ShowDetails.prototype, 'details', 'get').mockResolvedValue({
            id: 1,
            title: 'Title',
            genres: [],
            summary: 'Summary',
            type: 'Type',
            language: 'Language',
            website: 'website',
            episodes: []
        })
        const push = vi.spyOn(router, 'push')
        const wrapper = mount(ShowDetailsView)

        await flushPromises();
        
        (wrapper.vm as any).useCase.emit('showDetails:return')

        expect(push).toHaveBeenCalledWith('/')
    })
})