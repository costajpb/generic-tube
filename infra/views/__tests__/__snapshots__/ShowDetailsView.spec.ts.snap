// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShowDetailsView > should render 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="_container_9b1f4b"
  >
    <div>
      <div
        class="_skeleton_587f33 _title_9b1f4b"
      />
    </div>
    <div>
      <div
        class="_skeleton_587f33 _content_9b1f4b"
      />
    </div>
  </div>
  <!--v-if-->
  
</div>
`;
